src/tasks/task-angular-momentum-equality.cpp

```diff
@@ -0,0 +1,145 @@
+// tsid If not, see
+// <http://www.gnu.org/licenses/>.
+//
+
+#include "tsid/tasks/task-angular-momentum-equality.hpp"
+#include "tsid/robots/robot-wrapper.hpp"
+#include <pinocchio/algorithm/joint-configuration.hpp>
+#include <pinocchio/algorithm/centroidal.hpp>
+
+namespace tsid
+{
+  namespace tasks
+  {
+    using namespace math;
+    using namespace trajectories;
+    using namespace pinocchio;
+
+    TaskAMEquality::TaskAMEquality(const std::string & name,
+                                     RobotWrapper & robot):
+      TaskMotion(name, robot),
+      m_constraint(name, 3, robot.nv())
+    {
+      m_Kp.setZero(3);
+      m_Kd.setZero(3);
+      m_L_error.setZero(3);
+      m_dL_error.setZero(3);
+      m_L.setZero(3);
+      m_dL.setZero(3);
+      m_dL_des.setZero(3);
+    }
+
+    int TaskAMEquality::dim() const
+    {
+      //return self._mask.sum ()
+      return 3;
+    }
+
+    const Vector3 & TaskAMEquality::Kp(){ return m_Kp; }
+
+    const Vector3 & TaskAMEquality::Kd(){ return m_Kd; }
+
+    void TaskAMEquality::Kp(ConstRefVector Kp)
+    {
+      assert(Kp.size()==3);
+      m_Kp = Kp;
+    }
+
+    void TaskAMEquality::Kd(ConstRefVector Kd)
+    {
+      assert(Kd.size()==3);
+      m_Kd = Kd;
+    }
+
+    void TaskAMEquality::setReference(const TrajectorySample & ref)
+    {
+      m_ref = ref;
+    }
+
+    const TrajectorySample & TaskAMEquality::getReference() const
+    {
+      return m_ref;
+    }
+
+    const Vector3 & TaskAMEquality::getDesiredMomentumDerivative() const
+    {
+      return m_dL_des;
+    }
+
+    Vector3 TaskAMEquality::getdMomentum(ConstRefVector dv) const
+    {
+      return m_constraint.matrix()*dv - m_drift;
+    }
+
+    const Vector3 & TaskAMEquality::momentum_error() const
+    {
+      return m_L_error;
+    }
+
+    const Vector3 & TaskAMEquality::momentum() const
+    {
+      return m_L;
+    }
+    const Vector & TaskAMEquality::momentum_ref() const
+    {
+      return m_ref.vel;
+    }
+
+    const Vector & TaskAMEquality::dmomentum_ref() const
+    {
+      return m_ref.acc;
+    }
+
+    const ConstraintBase & TaskAMEquality::getConstraint() const
+    {
+      return m_constraint;
+    }
+
+    const ConstraintBase & TaskAMEquality::compute(const double ,
+                                                    ConstRefVector q,
+                                                    ConstRefVector v,
+                                                    const Data & data)
+    {
+      // Compute errors
+      // Get momentum jacobian
+      const Matrix6x & J_am = m_robot.momentumJacobian(data);
+      m_L = J_am.bottomRows(3) * v;
+      m_L_error = m_L - m_ref.vel;
+
+      m_dL_des = - m_Kp.cwiseProduct(m_L_error)
+                + m_ref.acc;
+
+#ifndef NDEBUG
+//      std::cout<<m_name<<" errors: "<<m_L_error.norm()<<" "
+//        <<m_dL_error.norm()<<std::endl;
+#endif
+      // Del Prete's quick and dirty way to compute drift
```

Comment:
You can probably use [this function](https://gepettoweb.laas.fr/doc/stack-of-tasks/pinocchio/master/doxygen-html/namespacepinocchio.html#a58aae3c6002a58f93c16b80a81c2e9bb) to compute the rate of change of the (angular) momentum with zero joint accelerations, which is the `drift`. You can use the "quick&dirty" computation in a test to check that what you are computing is indeed correct.

Reply:
I did it in my latests commits. The difference between the results of the two methods are between 1e-3 and 4e-3. 

Should I add this method to the RobotWrapper class ?

Reply:
Yes please, you can add it to RobotWrapper and then you're good to go for me! :)

Reply:
@pFernbach I think you are not doing the finite differences in the correct way. You should take into account the displacement due to the motion of the frame where the quantity is expressed.

Reply:
@andreadelprete I pushed the changes

@jcarpent Ok that explain the small differences between this and the `pinocchio::computeCentroidalMomentumTimeVariation` method. Anyway I am now using the method from pinocchio and I removed this finite difference code.

Reply:
> @pFernbach I think you are not doing the finite differences in the correct way. You should take into account the displacement due to the motion of the frame where the quantity is expressed.

Agreed. That's why the comment said "quick & dirty"! :)